### ISSUE TITLE
Remove object from array

### ISSUE THREAD

**Issue by ReneHollander** [svaarala/duktape#266]
I want to remove an object at a given index from an array. I am using the following code to delete the last item:
```c
// input: [{"key":"value"}]
duk_del_prop_index(ctx, -1, arrayLength - 1); 
// output: [null]
```
But it actually resets it with `null`. I would expect that the index is removed and the array getting shorter by one item.

**Comment by svaarala** [comment]
In Javascript array deletions don't affect .length.  For example:

```
$ node
> a = [1,2,3]
[ 1, 2, 3 ]
> delete a[2]
true
> a.length
3
> a
[ 1, 2,  ]
```

**Comment by ReneHollander** [comment]
Shame on me. Now I remember why I am always using a custom remove method with Array.splice to achieve the effect...
 Is there something similar I can do with the C-API?

**Comment by svaarala** [comment]
Right now there are unfortunately no array bindings in the C API e.g. to do operations like slice().

**Comment by ReneHollander** [comment]
Okay, thank you for your quick response! I will work around the issue for now :)

**Comment by svaarala** [comment]
You're welcome :) Let's leave this issue open because array C API bindings are a clear area for improvement.

**Comment by fatcerberus** [comment]
Wait, it resets the value to `null`?  Are you sure?  Because if so that would be a bug--`delete` (and therefore `duk_del_prop_index()`) should reset the property to `undefined`.

**Comment by svaarala** [comment]
Yes, it deletes the property; which is different to setting to undefined:

```
duk> a = [1,2,3]
= 1,2,3
duk> delete a[0]; a[1] = undefined; a[2] = null
= null
duk> print(a[0])
undefined
= undefined
duk> print(a[1])
undefined
= undefined
duk> print(a[2])
null
= undefined
duk> print('0' in a, '1' in a, '2' in a)
false true true
= undefined

```

**Comment by fatcerberus** [comment]
Oh, I knew that. I already know all about JS's idiosyncrasies regarding 'undefined`. :)  I just wanted to make sure the property wasn't actually evaluating as `null` after deletion because that would definitely be a bug.

**Comment by ReneHollander** [comment]
I am also missing a `duk_prop_insert_index(ctx, -2, 5);` to insert a value at a index into an array.

**Comment by fatcerberus** [comment]
As a workaround, you can do stuff like this by pushing the array onto the stack and then calling .splice on it yourself:

```c
// let's assume the array is on top of the stack already...

duk_get_prop_string(ctx, -1, "splice");  // load arr.splice onto stack
duk_dup(ctx, -2);  // this
duk_push_int(ctx, 5);  // arg1: start
duk_push_int(ctx, 0);  // arg2: deleteCount
duk_push_int(ctx, 812);  // item1
duk_call_method(ctx, 3);  // nargs = 3
duk_pop(ctx);  // retval
```

Which will insert the value `812` at index 5 of the array.

**Comment by ReneHollander** [comment]
Thank you!

I am curious on how big the performance impact is with workarounds like this to get missing functionality  into the C API. I would think, because the splice function is implemented in C, it is not too bad...

**Comment by fatcerberus** [comment]
In theory the only performance hit compared to a native API would be the property lookup.  Duktape's implementation is very tight and the APIs that *are* exposed generally run the exact same code as when the same constructs are used from script.

**Comment by ReneHollander** [comment]
Good to know!

And also thank you for your quick help and responses.

**Comment by svaarala** [comment]
@ReneHollander The snippet @fatcerberus posted is pretty much the best approach now. In fact there isn't any faster mechanism internally right now for most operations. For example, for splice() there is the Duktape/C function providing `Array.prototype.splice` which would be called as a method (via the value stack) anyway.

One improvement would be to provide pushers for the built-in methods so that you didn't need to look them up using a string. This would be footprint neutral but a bit faster. It could also provide the "original" method in case the method had been replaced later.

**Comment by ReneHollander** [comment]
A way to get the actual function Duktape implements would be nice, because when somebody modifies `Array.prototype.splice` it could, in my case at least, bring the whole application to a halt or at least cause some weird behavior.
Would it be a workaround for now, to save the function to the heap stash at startup and use it later when needed to overcome the issue or would the modification of the prototype also override the function in the heap stash? (Sounds weird, but you never know. I am still figuring out how Duktape does some of the things :D)

**Comment by svaarala** [comment]
Equally one might in other cases want the overridden behavior - it's presumably overridden for a reason :)

Anyway, if you store a function reference to the heap stash, that won't get modified if Array.prototype is modified - it would work the same as the Ecmascript code:

```js
    var origSplice = Array.prototype.splice;
    Array.prototype.splice = replacementSplice;

    // origSplice is still the original function
```

**Comment by ReneHollander** [comment]
Thanks!

Btw: I don't want to open another issue, but I am currently developing a object oriented Java Binding for Duktape: https://github.com/ReneHollander/duktape-java
I try to hide as much of the stack operations behind Java Objects for easier use. Duktape objects and arrays can be accessed from Java like maps and lists. In the near future it should also be possible to call Java methods from javascript and vice versa.
Maybe you want it to add it to the list on the website.

**Comment by svaarala** [comment]
Ok, will do!
